test/GPv2Settlement.e2e.ts

```diff
@@ -0,0 +1,201 @@
+        buyToken: eur.address,
+        sellAmount: ethers.utils.parseEther("15.0"),
+        buyAmount: ethers.utils.parseEther("180.0"),
+        appData: 4,
+      },
+      ethers.constants.Zero,
+    );
+
+    await addOrder(
+      traders[2],
+      {
+        ...orderDefaults,
+        kind: OrderKind.BUY,
+        partiallyFillable: true,
+        buyToken: oil.address,
+        sellToken: eur.address,
+        buyAmount: ethers.utils.parseEther("4.0"),
+        sellAmount: ethers.utils.parseEther("52.0"),
+        appData: 5,
+      },
+      ethers.utils.parseEther("27.0").div(13),
+    );
+
+    await addOrder(
+      traders[3],
+      {
+        ...orderDefaults,
+        kind: OrderKind.BUY,
+        partiallyFillable: true,
+        buyToken: wine.address,
+        sellToken: eur.address,
+        buyAmount: ethers.utils.parseEther("20.0"),
+        sellAmount: ethers.utils.parseEther("280.0"),
+        appData: 6,
+      },
+      ethers.utils.parseEther("12.0"),
+    );
+
+    await settlement.connect(solver).settle(
+      encoder.tokens,
+      encoder.clearingPrices({
+        [eur.address]: ethers.utils.parseEther("1.0"),
+        [oil.address]: ethers.utils.parseEther("13.0"),
+        [wine.address]: ethers.utils.parseEther("14.0"),
+      }),
+      encoder.encodedTrades,
+      "0x",
+      "0x",
+    );
+
+    expect(await wine.balanceOf(traders[0].address)).to.deep.equal(
+      STARTING_BALANCE.sub(ethers.utils.parseEther("12.0")).sub(
+        orderDefaults.feeAmount,
+      ),
+    );
+    expect(await oil.balanceOf(traders[0].address)).to.deep.equal(
+      ethers.utils.parseEther("12.0").mul(14).div(13),
+    );
+
+    expect(await oil.balanceOf(traders[1].address)).to.deep.equal(
+      STARTING_BALANCE.sub(ethers.utils.parseEther("15.0")).sub(
+        orderDefaults.feeAmount,
+      ),
+    );
+    expect(await eur.balanceOf(traders[1].address)).to.deep.equal(
+      ethers.utils.parseEther("15.0").mul(13),
+    );
+
+    expect(await eur.balanceOf(traders[2].address)).to.deep.equal(
+      STARTING_BALANCE.sub(ethers.utils.parseEther("27.0"))
+        .sub(
+          orderDefaults.feeAmount
+            .mul(ethers.utils.parseEther("27.0").div(13))
+            .div(ethers.utils.parseEther("4.0")),
+        )
+        // NOTE: Account for rounding error from computing sell amount that is
+        // an order of magnitude larger than executed buy amount from the
+        // settlement.
+        .add(1),
+    );
+    expect(await oil.balanceOf(traders[2].address)).to.deep.equal(
+      ethers.utils.parseEther("27.0").div(13),
+    );
+
+    expect(await eur.balanceOf(traders[3].address)).to.deep.equal(
+      STARTING_BALANCE.sub(ethers.utils.parseEther("12.0").mul(14)).sub(
+        orderDefaults.feeAmount
+          .mul(ethers.utils.parseEther("12.0"))
+          .div(ethers.utils.parseEther("20.0")),
+      ),
+    );
+    expect(await wine.balanceOf(traders[3].address)).to.deep.equal(
```

Comment:
Why is wine balance only checked for two addresses where as eur and oil is checked for three? Also, should we assert the remaining balances in the settlement contract (the fee)?

Reply:
We just check the buy and sell token balances for each trader. So:
- trader 0 sells wine sells oil
- trader 1 sells oil buys eur
- trader 2 buys oil sell eur
- trader 3 buys wine sells eur.

So that 2 wine, 3 oil and 3 eur. We can also assert that the other token balance is zero if you think that has value?

Reply:
I think it's because wine only appears in two orders, while all others appear in three. But it's weird, maybe a diagram with all orders would help.

Reply:
Added!